src/Take.Elephant.ElasticSearch/StorageBase.cs

```diff
@@ -0,0 +1,131 @@
+            Expression<Func<T, bool>> where,
+            Expression<Func<T, TResult>> select, int skip, int take,
+            CancellationToken cancellationToken = default)
+        {
+            var queryDescriptor = where.ParseToQueryContainer<T>();
+
+            var result = await ElasticClient.SearchAsync<T>(s => s
+                .Index(_index)
+                .Query(_ => queryDescriptor)
+                .From(skip).Size(take), cancellationToken);
+
+            return new QueryResult<T>(new AsyncEnumerableWrapper<T>(result.Documents), (int)result.Total);
+        }
+
+        public async Task<bool> ContainsKeyAsync(string key, CancellationToken cancellationToken = default)
+        {
+            if (key == null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            var response = await ElasticClient.DocumentExistsAsync<T>(key, d => d
+                .Index(_index), cancellationToken);
+
+            return response.Exists;
+        }
+
+        public async Task<T> GetValueOrDefaultAsync(string key, CancellationToken cancellationToken = default)
+        {
+            if (key == null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            var result = await ElasticClient.SearchAsync<T>(s => s
+                .Index(_index)
+                .Query(q => q.Ids(a => a.Values(key.ToString()))), cancellationToken);
+
+            return result.Documents.FirstOrDefault();
+        }
+
+        public async Task<bool> TryAddAsync(string key, T value, bool overwrite = false, CancellationToken cancellationToken = default)
+        {
+            if (key == null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            if (value == null)
+            {
+                throw new ArgumentNullException(nameof(value));
+            }
+
+            if (overwrite || !await ContainsKeyAsync(key, cancellationToken))
+            {
+                var result = await ElasticClient.IndexAsync(new IndexRequest<T>(value, _index,
+                    key.ToString()), cancellationToken);
+
+                return result.IsValid;
+            }
+
+            return false;
+        }
+
+        public async Task<bool> DeleteAsync(string key, CancellationToken cancellationToken = default)
+        {
+            if (key == null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            var result = await ElasticClient.DeleteAsync<T>(key,
+                d => d.Index(_index), cancellationToken);
+
+            return result.IsValid;
+        }
+
+        protected string GetPropertyValue(T entity, string property)
+        {
+            if (property == null)
+            {
+                throw new ArgumentNullException(property);
+            }
+
+            return entity.GetType().GetProperties()
```

Comment:
This reflection call is slow. You can use a cached delegate instead, like is done in the TypeMapper type.

Look here: https://github.com/takenet/elephant/blob/master/src/Take.Elephant.Sql/Mapping/TypeMapper.cs#L44

Create an accessor and save it in a concurrent dictionary (use the GetOrAdd method)